// weapp-tailwindcss-webpack-plugin/dist/v4.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('./index-62caf23f.js');
var index$1 = require('./index-33e84d7e.js');
require('./defaults-464f242a.js');
require('./utils-17c91589.js');
require('micromatch');
require('@babel/generator');
require('@babel/parser');
require('@babel/traverse');
require('postcss');
require('postcss-selector-parser');
require('path');
require('fs');
require('semver');
require('./constants-ff495628.js');
require('webpack-sources');
require('./shared-6c9547c3.js');
require('webpack');



exports.BaseJsxWebpackPluginV4 = index.BaseJsxWebpackPluginV4;
exports.BaseTemplateWebpackPluginV4 = index.BaseTemplateWebpackPluginV4;
exports.TaroWeappTailwindcssWebpackPluginV4 = index.TaroWeappTailwindcssWebpackPluginV4;
exports.UniAppWeappTailwindcssWebpackPluginV4 = index.UniAppWeappTailwindcssWebpackPluginV4;
exports.KboneWeappTailwindcssWebpackPluginV4 = index$1.KboneWeappTailwindcssWebpackPluginV4;
exports.RemaxWeappTailwindcssWebpackPluginV4 = index$1.RemaxWeappTailwindcssWebpackPluginV4;
