// weapp-tailwindcss-webpack-plugin/dist/vite.js
'use strict';

var defaults = require('./defaults-464f242a.js');
var constants = require('./constants-ff495628.js');
var shared = require('./shared-6c9547c3.js');
require('./utils-17c91589.js');
require('micromatch');
require('@babel/generator');
require('@babel/parser');
require('@babel/traverse');
require('postcss');
require('postcss-selector-parser');
require('path');
require('fs');
require('semver');

function ViteWeappTailwindcssPlugin(options = {}) {
    const opts = defaults.getOptions(options, ['patch', 'style', 'templete']);
    const { disabled, onEnd, onLoad, onStart, onUpdate, templeteHandler, styleHandler, patch } = opts;
    if (disabled) {
        return;
    }
    patch === null || patch === void 0 ? void 0 : patch();
    onLoad();
    return {
        name: constants.vitePluginName,
        enforce: 'post',
        buildStart() {
            onStart();
        },
        generateBundle(opt, bundle, isWrite) {
            const entries = Object.entries(bundle).filter(([, s]) => s.type === 'asset');
            const groupedEntries = shared.getGroupedEntries(entries, opts);
            if (Array.isArray(groupedEntries.html)) {
                for (let i = 0; i < groupedEntries.html.length; i++) {
                    const [file, originalSource] = groupedEntries.html[i];
                    const oldVal = originalSource.source.toString();
                    originalSource.source = templeteHandler(oldVal);
                    onUpdate(file, oldVal, originalSource.source);
                }
            }
            if (Array.isArray(groupedEntries.css)) {
                for (let i = 0; i < groupedEntries.css.length; i++) {
                    const [file, originalSource] = groupedEntries.css[i];
                    const rawSource = originalSource.source.toString();
                    const css = styleHandler(rawSource, {
                        isMainChunk: true
                    });
                    originalSource.source = css;
                    onUpdate(file, rawSource, css);
                }
            }
        },
        buildEnd() {
            onEnd();
        }
    };
}

module.exports = ViteWeappTailwindcssPlugin;
